json/src/fs2/data/json/SelectorParser.scala

```diff
@@ -86,15 +86,21 @@ class SelectorParser[F[_]](val input: String)(implicit F: MonadError[F, Throwabl
   private def parseArraySelector(idx: Int, fst: Int): F[(Boolean => Selector, Int)] =
     next(idx).flatMap {
       case Some((RightBracket(_), idx)) => F.pure((Selector.IndexSelector(fst, _), idx))
-      case Some((Comma(_), _))          => parseSeveralIndices(idx, Seq(fst))
+      case Some((Comma(_), _))          => parseSeveralIndices(idx, List(fst))
       case Some((Colon(_), idx))        => parseRangeIndices(idx, fst)
       case Some((token, _)) =>
         F.raiseError(new JsonSelectorException(s"invalid selector token '$token'", token.idx))
       case None =>
         F.raiseError(new JsonSelectorException("unexpected end of input", idx))
     }
 
-  private def parseSeveralNames(idx: Int, names: Seq[String]): F[(Boolean => Selector, Int)] =
+  private def parseMandatory(idx: Int, names: List[String]): F[(Boolean => Selector, Int)] =
+    next(idx, peek = true).flatMap {
+      case Some((Bang(_), idx)) => F.pure((Selector.NameSelector(names, _, true), idx + 1))
+      case _ => F.pure((Selector.NameSelector(names, _, false), idx))
+    }
```

Comment:
No need for `flatMap` as both cases call `F.pure`:
```suggestion
    next(idx, peek = true).map {
      case Some((Bang(_), idx)) => (Selector.NameSelector(names, _, true), idx + 1)
      case _ => (Selector.NameSelector(names, _, false), idx)
    }
```

Reply:
Indeed.